### PR TITLE
add material slider secondary value

### LINKED ISSUE #64029: Add secondaryValue to Slider
## Use case

Audio/video players need to display a seek bar, often with a secondary value indicating the buffered position. This allows the user to see how much has downloaded in a different color to how much has played. The Android equivalent of Slider (ProgressBar) supports this via [secondaryProgress](https://developer.android.com/reference/android/widget/ProgressBar#attr_android:secondaryProgress).

There is a flutter plugin that does this called [seekbar](https://pub.dev/packages/seekbar).

Another use case is a progress bar for downloading/installing or downloading/extracting a file where an app may want to show both the download progress and the extraction/installation progress, where one of the progress values is always ahead of the other.

## Proposal

I would propose adding a `secondaryValue` to Slider along with a `secondaryColor`.

One reason to add this into Slider rather than as a separate plugin on pub.dev is that any pub.dev package that does this would have 99% of its code in common with Slider and would not benefit when slider gains new features, and there is also precedent with Android's widget library having this feature built into its Slider equivalent.

### PR DESCRIPTION
**Pull request by akshdeep-singh**
This PR adds the support for showing secondary value in Slider widget.

Slider 

Fixes #64029

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 1]
```diff
@@ -1593,6 +1626,25 @@ class RectangularSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
     if (!rightTrackSegment.isEmpty) {
       context.canvas.drawRect(rightTrackSegment, rightTrackPaint);
     }
+
+    final bool showMiddleTrack = (secondaryOffset != null) &&
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/slider_theme.dart:1630]: Middle track?  It seems like you should stick to similar terminology, like `showSecondaryTrack`.

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: examples/api/lib/material/slider/slider.1.dart [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (gspencergoog) [inline examples/api/lib/material/slider/slider.1.dart:1]: You'll need to add a test for this example. Take a look at the other example tests for how to do them (they're pretty simple).

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Once you add a test for the example, this is good to go.  Thanks for adding this, it looks pretty cool!

You'll need a second review before we can merge this: I'll ask and see if anyone is available.

File: packages/flutter/lib/src/material/slider.dart [hunk 3]
```diff
@@ -182,13 +189,25 @@ class Slider extends StatefulWidget {
        assert(max != null),
        assert(min <= max),
        assert(value >= min && value <= max),
+       assert(secondaryTrackValue == null || (secondaryTrackValue >= min && secondaryTrackValue <= max)),
```
Comment (justinmc) [inline packages/flutter/lib/src/material/slider.dart:192]: Above you included a message with this assert, should you include it here too?  Though maybe it's obvious.

File: packages/flutter/lib/src/material/slider.dart [hunk 4]
```diff
@@ -182,13 +189,25 @@ class Slider extends StatefulWidget {
...
        assert(divisions == null || divisions > 0);
 
   /// The currently selected value for this slider.
   ///
   /// The slider's thumb is drawn at a position that corresponds to this value.
   final double value;
 
+  /// The secondary track value for this slider
```
Comment (justinmc) [inline packages/flutter/lib/src/material/slider.dart:200]: Period at the end of this sentence.

File: packages/flutter/lib/src/material/slider.dart [hunk 5]
```diff
@@ -365,6 +384,18 @@ class Slider extends StatefulWidget {
   /// Ignored if this slider is created with [Slider.adaptive].
   final Color? inactiveColor;
 
+  /// The color to use for the portion of the slider track between the thumb and
+  /// the [Slider.secondaryTrackValue]
```
Comment (justinmc) [inline packages/flutter/lib/src/material/slider.dart:388]: Period at the end here.

**Review by justinmc — APPROVED** [review]
LGTM with nits 👍

File: examples/api/lib/material/slider/slider.1.dart (hunk 2 above)
Comment (dkwingsmt) [reply examples/api/lib/material/slider/slider.1.dart:1]: I might have mistaken but can you point me wherer this example is used in docs?

File: packages/flutter/lib/src/material/slider.dart [hunk 6]
```diff
@@ -125,6 +125,7 @@ class Slider extends StatefulWidget {
   const Slider({
     super.key,
     required this.value,
+    this.secondaryTrackValue,
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/slider.dart:128]: It is unclear to me what a "secondary value" is. I suggest adding a paragraph explaining the secondary value, for example adapting from the issue this PR is fixing, possibly adding it to the doc of Slider or secondaryValue.

File: examples/api/lib/material/slider/slider.1.dart (hunk 2 above)
Comment (dkwingsmt) [inline examples/api/lib/material/slider/slider.1.dart:1]: Can you point to me where this example is used? I don't see it referred to anywhere in doc.

File: packages/flutter/lib/src/material/slider.dart (hunk 6 above)
Comment (akshdeep-singh) [reply packages/flutter/lib/src/material/slider.dart:128]: I think I have already added information at line 209. Am I missing something here?

File: examples/api/lib/material/slider/slider.1.dart (hunk 2 above)
Comment (akshdeep-singh) [reply examples/api/lib/material/slider/slider.1.dart:1]: Add in latest commit

Comment (akshdeep-singh) [reply examples/api/lib/material/slider/slider.1.dart:1]: added in latest commit

File: packages/flutter/lib/src/material/slider.dart (hunk 3 above)
Comment (akshdeep-singh) [reply packages/flutter/lib/src/material/slider.dart]: done

**Comment by akshdeep-singh** [comment]
@gspencergoog I think everything is done now? It can be merged then!

**Pull request merged**